### ISSUE TITLE
transaction on_rollback triggers are called from a foreign fiber when the roll back is triggered by other sync tx timeout

### ISSUE THREAD

**Issue by sergepetrenko** [tarantool/tarantool#7908]
When there is a pending synchronous transaction, all other transactions (even async ones) enter the syncrhonous transaction queue and wait for the sync transaction completion. If that sync transaction is then rolled back due to timeout, it rolls back all the other transactions in the queue, running their on_rollback triggers in the sync transaction fiber (not their own fibers).

This is bad, for example, when the user expects to use `fiber.storage` inside the on_rollback triggers.

Here's the <details><summary>reproducer</summary>
```
tarantool> box.cfg{replication_synchro_quorum=2, replication_synchro_timeout=1}
---
...

tarantool> box.schema.space.create('sync', {is_sync=true}):create_index('pk')
---
...

tarantool> box.schema.space.create("async"):create_index('pk')
---
...

tarantool> log = require'log'
---
...

tarantool> fiber =require('fiber')
---
...

tarantool> box.ctl.promote()
---
...

tarantool> fiber.create(function() box.space.sync:replace{1} end) box.begin() box.on_rollback(function() log.info("Rollback from that fiber") end) log.info("Transaction in this fiber") box.space.async:replace{1} box.commit()
2022-11-08 12:22:30.430 [18722] main/103/interactive I> Transaction in this fiber
2022-11-08 12:22:31.431 [18722] main/114/lua I> Rollback from that fiber
2022-11-08 12:22:31.431 [18722] main/114/lua txn_limbo.c:295 E> ER_SYNC_QUORUM_TIMEOUT: Quorum collection for a synchronous transaction is timed out
---
- error: A rollback for a synchronous transaction is received
...

```

</details>

Note the
```
2022-11-08 12:22:30.430 [18722] main/103/interactive I> Transaction in this fiber
2022-11-08 12:22:31.431 [18722] main/114/lua I> Rollback from that fiber
```
lines. The fiber ids of fibers are different: 103 and 114. Meaning that transaction on_rollback triggers are executed in some other fiber (not the one the transaction ran in).

**Comment by Gerold103** [comment]
I would note that this won't work for fiber-less transactions anyway. Like those coming from applier. I would rather advice people not rely on the end-txn triggers running in their fiber. On-commit triggers don't run in the same fiber even for the blocking txns. They are run immediately when commit happens.

I would rather expect rollback triggers to be consistent and do the same - work exactly when rollback happens, immediately. I personally have some logic on that in vshard, when I expect to update certain global things immediately, before the committed or rolled back data could be accessed by anybody. See https://github.com/tarantool/vshard/blob/master/vshard/storage/init.lua#L433 (`bucket_on_commit_or_rollback_f()`).

As a proof of the on-commit case, see the example:
```Lua
box.cfg{}
fiber = require('fiber')
log = require('log')

s = box.schema.create_space('test')
_ = s:create_index('pk')

log.info('txn fid: %s', fiber.self():id())
do
    box.begin()
    box.on_commit(function()
        log.info('commit fiber: %s', fiber.self())
    end)
    s:replace{1}
    box.commit()
end
```
The commit trigger here prints `nil`. It is not just a different fiber. It is not a public fiber at all. This is the scheduler fiber.

TLDR: I don't think it is a bug. The issue should be closed and maybe the behaviour must be more explicitly documented, IMO.

**Comment by sergepetrenko** [comment]
@Gerold103, thanks for the explanation! Sounds legit.

**Comment by alyapunov** [comment]
Thanks for your investigation!
